api/buckets/service.go

```diff
@@ -403,13 +492,14 @@ func (nn *namedNodes) Store(c cid.Cid, n *namedNode) {
 // Leaf nodes are encrypted and linked to parents, which are then encrypted and
 // linked to their parents, and so on up to root.
 // add will be added to the encrypted root node if not nil.
-// This method returns a map of all nodes keyed by their _original_ plaintext cid.
-func (s *Service) encryptDag(ctx context.Context, ds ipld.DAGService, root ipld.Node, key []byte, add *namedNode) (map[cid.Cid]*namedNode, error) {
+// This method returns a map of all nodes keyed by their _original_ plaintext cid,
+// and a list of the root's direct links.
+func (s *Service) encryptDag(ctx context.Context, ds ipld.DAGService, root ipld.Node, destPath string, linkKey []byte, currentFileKeys, newFileKeys map[string][]byte, newFileKey []byte, add *namedNode) (map[cid.Cid]*namedNode, error) {
```

Comment:
This method can now decrypt the dag before re-encrypting with the given key map. The key map holds the correct file key for each path, unless a single key is given for all paths (the case when creating a bucket from an existing Cid).